types/lodash/ts3.1/common/array.d.ts

```diff
@@ -0,0 +1,2120 @@
+         */
+        findLastIndex<T>(array: List<T> | null | undefined, predicate?: ListIterateeCustom<T, boolean>, fromIndex?: number): number;
+    }
+    interface Collection<T> {
+        /**
+         * @see _.findLastIndex
+         */
+        findLastIndex(predicate?: ListIterateeCustom<T, boolean>, fromIndex?: number): number;
+    }
+    interface CollectionChain<T> {
+        /**
+         * @see _.findLastIndex
+         */
+        findLastIndex(predicate?: ListIterateeCustom<T, boolean>, fromIndex?: number): PrimitiveChain<number>;
+    }
+    interface LoDashStatic {
+        /**
+         * @see _.head
+         */
+        first: LoDashStatic["head"];
+    }
+    interface String {
+        /**
+         * @see _.first
+         */
+        first(): string | undefined;
+    }
+    interface StringChain {
+        /**
+         * @see _.first
+         */
+        first(): StringNullableChain;
+    }
+    interface StringNullableChain {
+        /**
+         * @see _.first
+         */
+        first(): StringNullableChain;
+    }
+    interface Collection<T> {
+        /**
+         * @see _.first
+         */
+        first(): T | undefined;
+    }
+    interface CollectionChain<T> {
+        /**
+         * @see _.first
+         */
+        first(): ExpChain<T | undefined>;
+    }
+    interface RecursiveArray<T> extends Array<T|RecursiveArray<T>> {}
+    interface ListOfRecursiveArraysOrValues<T> extends List<T|RecursiveArray<T>> {}
+    interface LoDashStatic {
+        /**
+         * Flattens `array` a single level deep.
+         *
+         * @param array The array to flatten.
+         * @return Returns the new flattened array.
+         */
+        flatten<T>(array: List<Many<T>> | null | undefined): T[];
+    }
+    interface String {
+        /**
+         * @see _.flatten
+         */
+        flatten(): Collection<string>;
+    }
+    interface StringChain {
+        /**
+         * @see _.flatten
+         */
+        flatten(): CollectionChain<string>;
+    }
+    interface StringNullableChain {
+        /**
+         * @see _.flatten
+         */
+        flatten(): CollectionChain<string>;
+    }
+    interface Collection<T> {
+        /**
+         * @see _.flatten
+         */
+        flatten(): T extends Many<infer U> ? Collection<U> : Collection<T>;
+    }
+    interface CollectionChain<T> {
+        /**
+         * @see _.flatten
+         */
+        flatten(): T extends Many<infer U> ? CollectionChain<U> : CollectionChain<T>;
+    }
+    interface LoDashStatic {
+        /**
+         * Recursively flattens a nested array.
+         *
+         * @param array The array to recursively flatten.
+         * @return Returns the new flattened array.
+         */
+        flattenDeep<T>(array: ListOfRecursiveArraysOrValues<T> | null | undefined): T[];
+    }
+    interface Collection<T> {
+        /**
+         * @see _.flattenDeep
+         */
+        flattenDeep(): Collection<T>;
```

Comment:
Too bad we can't really get the right type here. I feel like `T` is almost always the wrong type though. Maybe `Collection<any>` would be better? Or maybe something like what you did for `flatMapDeep`:

```ts
flattenDeep<TResult = any>(): Collection<TResult>;
```

Same below.

Reply:
Well, I think the type will be a supertype of the correct type at least. Consider:

```ts
[1, [2, 3, [3, 4]], 6, 7]: (number | (number | number[])[])[]
```

flattenDeep of that should be `number[]` but `number[]` is a subtype of `(number | (number | number[])[])[]` at least, so you can manually narrow using casts probably. Type parameters are just an alternate syntax for a cast in this case, and I'd rather not encourage that. So I'll skip the type parameter here; `flatMapDeep` was an existing use of this pattern that I didn't want to break.